decouple service default loop from bootstrap controller

aojea opened the pull request:
Kubernetes cluster run a special service in the default namespace.

The "kuberntes.default" service is used for InCluster configuration and is maintained by the kube-apiservers in the cluster.

This commit decouples the service reconcilation logic from the bootstrap controller.

The kubernetes.default service is only reconciled if it doesn't exist, this may be unexpected but this has been the default behavior since https://github.com/kubernetes/kubernetes/commit/2a9a9fa155e92d0bd7f4eb85d843cfa14b830cca

/kind cleanup

```release-note
NONE
```

/sig api-machinery

aojea commented:
/assign @wojtek-t 

2/3 after https://github.com/kubernetes/kubernetes/pull/116170

cici37 commented:
/triage accepted

wojtek-t commented:
/remove-lifecycle stale

I'm planning to get to it after feature freeze

wojtek-t commented on pkg/controlplane/controller.go:
```diff
@@ -147,14 +91,7 @@ func (c *Controller) Start() {
-		klog.Errorf("Error removing old endpoints from kubernetes service: %v", err)
-	}
-
+	repairClusterIPs := servicecontroller.NewRepair(c.ServiceClusterIPInterval, c.client.CoreV1(), c.client.EventsV1(), &c.ServiceClusterIPRange, c.ServiceClusterIPRegistry, &c.SecondaryServiceClusterIPRange, c.SecondaryServiceClusterIPRegistry)
```
Currently we're using different repair loops based on the FG `MultiCIDRServiceAllocator` - why you're changing this?

wojtek-t commented on pkg/controlplane/controller/defaultservice/default_service_controller.go:
```diff
@@ -0,0 +1,337 @@
+	// we construct our own informer because we need such a small subset of the information available.
+	// The kubernetes.default service
+	serviceInformer := corev1informers.NewFilteredServiceInformer(client, metav1.NamespaceDefault, 12*time.Hour,
+		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
```
Given we're watching from a single namespace (effectively a single object), creating an index here seems like overkill to me.

wojtek-t commented on pkg/controlplane/controller/defaultservice/default_service_controller.go:
```diff
@@ -0,0 +1,337 @@
+
+	// Reconcile during first run removing itself until server is ready.
+	endpointPorts := createEndpointPortSpec(c.endpointPort, "https")
+	if err := c.endpointReconciler.RemoveEndpoints(serviceName, c.endpointIP, endpointPorts); err == nil {
```
note to myself: this is a newly added thing, but that seems desired

aojea replied:
it is not new https://github.com/kubernetes/kubernetes/blob/75889ecec5d30cbe1dcb6636d5334b21b4378e9c/pkg/controlplane/controller.go#L150-L156

wojtek-t commented on pkg/controlplane/controller/defaultservice/default_service_controller.go:
```diff
@@ -0,0 +1,337 @@
+	go c.serviceInformer.Run(stopCh)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.servicesSynced) {
+		return
```
utilruntime.HandleError(...)

wojtek-t commented on pkg/controlplane/controller/defaultservice/default_service_controller.go:
```diff
@@ -0,0 +1,337 @@
+	// wait to publish the endpoints until the process is ready so we don't receive traffic
+	wait.PollImmediateUntil(100*time.Millisecond, func() (bool, error) {
+		var code int
+		c.client.CoreV1().RESTClient().Get().AbsPath("/healthz").Do(context.Background()).StatusCode(&code)
```
It was /readyz - why you're switching to /healthz?

aojea replied:
typo

wojtek-t commented on pkg/controlplane/controller/defaultservice/default_service_controller.go:
```diff
@@ -0,0 +1,337 @@
+	}, stopCh)
+
+	// wait until the default namespace has been created
+	wait.PollImmediateUntil(100*time.Millisecond, func() (bool, error) {
```
Not sure if this approach is actually better - if someone deletes the namespace in the meantime (but after this runs), this will not catch it.

aojea replied:
yep

wojtek-t commented on pkg/controlplane/controller/defaultservice/default_service_controller_test.go:
```diff
@@ -0,0 +1,284 @@
+		expectCreate *v1.Service // nil means none expected
+	}{
+		{
+			testName: "service does not exist",
```
Can you add a testcase where service actually exists?

[or maybe two cases: when service is the one expected and when it's different one?]

wojtek-t commented on pkg/controlplane/controller/defaultservice/default_service_controller_test.go:
```diff
@@ -0,0 +1,284 @@
+					}
+				}
+			}
+			if test.expectCreate == nil && len(creates) > 1 {
```
>0 ?

wojtek-t commented on pkg/controlplane/controller/defaultservice/default_service_controller_test.go:
```diff
@@ -0,0 +1,284 @@
+}
+
+func TestNewController(t *testing.T) {
+
```
nit: remove empty line

wojtek-t commented on pkg/controlplane/controller/defaultservice/default_service_controller.go:
```diff
@@ -0,0 +1,337 @@
+	// Service exists; warn if reconciliation is necessary
+	// The Service kubernetes.default is never reconciled once created since
+	// https://github.com/kubernetes/kubernetes/commit/2a9a9fa155e92d0bd7f4eb85d843cfa14b830cca
+	// TODO: revisit if is necessary to change this behavior, add
```
hmm - before your change we were reconciling service only on start

But with your change we're never reconciling it - seems like a bug...

aojea replied:
I'm rolling this changes back

wojtek-t commented on pkg/controlplane/controller/defaultservice/default_service_controller_test.go:
```diff
@@ -0,0 +1,284 @@
+		})
+	}
+}
+
```
I'm seeing a bunch of testcases (incorrect port, incorrect type, etc.) in the old tests that you didn't copy - why? These seems useful..

wojtek-t commented on pkg/controlplane/instance.go:
```diff
@@ -596,21 +596,45 @@ func (m *Instance) InstallLegacyAPI(c *completedConfig, restOptionsGetter generi
+		return fmt.Errorf("error creating kubernetes.default service controller: %v", err)
+	}
+	m.GenericAPIServer.AddPostStartHookOrDie("start-default-service-controller", defaultServiceController.PostStartHook)
+	m.GenericAPIServer.AddPreShutdownHookOrDie("start-default-service-controller", defaultServiceController.PreShutdownHook)
```
s/start/stop/   ?

wojtek-t commented on test/integration/controlplane/kube_apiserver_test.go:
```diff
@@ -606,3 +608,71 @@ func TestMultiAPIServerNodePortAllocation(t *testing.T) {
+
+		return !reflect.DeepEqual(svc.Spec, defaultService.Spec), nil
+	}); err != nil {
+		t.Fatalf("reconciling kubernetes service timed out")
```
I'm not following this - shouldn't it actually fail?

aojea replied:
if the services are equal it returns successfully

wojtek-t reviewed (commented):
Finally got to it - PTAL

aojea committed:
decouple service default loop from bootstrap controller

Kubernetes cluster run a special service in the default namespace.

The "kuberntes.default" service is used for InCluster configuration
and is maintained by the kube-apiservers in the cluster.

This commit decouples the service reconcilation logic from the
bootstrap controller.

The kubernetes.default service is only reconciled if it doesn't
exist, this may be unexpected but this has been the default behavior
since 2a9a9fa

aojea committed:
use informer instead of get

Change-Id: I39f5be2e046c8114d431bff21b905c5e1cd267ac

aojea commented:
/hold cancel

@wojtek-t @sttts modified the PR to be easier to review, there are two steps:
1. new controller with exactly the same logic , I just copied the methods directly
2. replace the periodic Get by the informer Get

aojea commented:
/hold

let's have only one, superseded by @sttts https://github.com/kubernetes/kubernetes/pull/119064
